source/ref_gl/r_vbo.c

```diff
@@ -57,48 +52,134 @@ static int r_num_active_vbos;
 static elem_t *R_VBOElemBuffer( unsigned numElems );
 static void *R_VBOVertBuffer( unsigned numVerts, size_t vertSize );
 
-/*
-* R_InitVBO
-*/
-void R_InitVBO( void )
-{
-	int i;
 
-	r_vbo_tempelems = NULL;
-	r_vbo_numtempelems = 0;
+mesh_vbo_t *R_CreateMeshVBO_NIRA( void *owner, int numVerts, int numElems, int numInstances, vattribmask_t vattribs, vbo_tag_t tag, vattribmask_t halfFloatVattribs ) {
+	size_t size;
+	vbohandle_t *vboh = NULL;
+	mesh_vbo_t *vbo = NULL;
+	GLenum usage = VBO_USAGE_FOR_TAG( tag );
+	vattribbit_t lmattrbit;
+	if( !( halfFloatVattribs & VATTRIB_POSITION_BIT ) ) {
+		halfFloatVattribs &= ~( VATTRIB_AUTOSPRITE_BIT );
+	}
 
-	r_vbo_tempvsoup = NULL;
-	r_vbo_tempvsoupsize = 0;
+	halfFloatVattribs &= ~VATTRIB_COLORS_BITS;
+	halfFloatVattribs &= ~VATTRIB_BONES_BITS;
 
-	r_num_active_vbos = 0;
+	// TODO: convert quaternion component of instance_t to half-float
+	// when uploading instances data
+	halfFloatVattribs &= ~VATTRIB_INSTANCES_BITS;
 
-	memset( r_mesh_vbo, 0, sizeof( r_mesh_vbo ) );
-	memset( r_vbohandles, 0, sizeof( r_vbohandles ) );
+	vboh = r_free_vbohandles;
+	vbo = &r_mesh_vbo[vboh->index];
 
-	// link vbo handles
-	r_free_vbohandles = r_vbohandles;
-	r_vbohandles_headnode.prev = &r_vbohandles_headnode;
-	r_vbohandles_headnode.next = &r_vbohandles_headnode;
-	for( i = 0; i < MAX_MESH_VERTEX_BUFFER_OBJECTS; i++ ) {
-		r_vbohandles[i].index = i;
-		r_vbohandles[i].vbo = &r_mesh_vbo[i];
+	size_t vertexSize = 0;
+
```

Comment:
probably just copy all the code since I expect this to diverge more later rather not touch the original code too much in this pass. this can be revisited later.